internal/primitives/runtime/generic/header.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package generic
+
+import (
+	"io"
+
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+// Header is a block header, and implements a compatible encoding to `sp_runtime::generic::Header`
+type Header[N runtime.Number, H runtime.Hash, Hasher runtime.Hasher[H]] struct {
+	// The parent hash.
+	parentHash H
+	// The block number.
+	number N
+	// The state trie merkle root
+	stateRoot H
+	// The merkle root of the extrinsics.
+	extrinsicsRoot H
+	// A chain-specific digest of data useful for light clients or referencing auxiliary data.
+	digest runtime.Digest
+}
+
+// Number returns the block number.
+func (h Header[N, H, Hasher]) Number() N {
+	return h.number
+}
+
+// SetNumber sets the block number.
+func (h *Header[N, H, Hasher]) SetNumber(number N) {
+	h.number = number
+}
+
+// ExtrinsicsRoot returns the extrinsics root.
+func (h Header[N, H, Hasher]) ExtrinsicsRoot() H {
+	return h.extrinsicsRoot
+}
+
+// SetExtrinsicsRoot sets the extrinsics root.
+func (h *Header[N, H, Hasher]) SetExtrinsicsRoot(root H) {
+	h.extrinsicsRoot = root
+}
+
+// StateRoot returns the state root.
+func (h Header[N, H, Hasher]) StateRoot() H {
+	return h.stateRoot
+}
+
+// SetStateRoot sets the state root.
+func (h *Header[N, H, Hasher]) SetStateRoot(root H) {
+	h.stateRoot = root
+}
+
+// ParentHash returns the parent hash.
+func (h Header[N, H, Hasher]) ParentHash() H {
+	return h.parentHash
+}
+
+// SetParentHash sets the parent hash.
+func (h *Header[N, H, Hasher]) SetParentHash(hash H) {
+	h.parentHash = hash
+}
+
+// Digest returns the digest.
+func (h Header[N, H, Hasher]) Digest() runtime.Digest {
+	return h.digest
+}
+
+// DigestMut returns a mutable reference to the stored digest.
+func (h Header[N, H, Hasher]) DigestMut() *runtime.Digest {
+	return &h.digest
+}
+
+type helper[H any] struct {
```

Comment:
Can we remove the `helper` and use the method `UnmarshalSCALE` in the `Header`? I really dislike we should maintain 2 structs (adds complexity), so looks like we should not be generic over the number type.

Also the getters/setters are really useful? Looks like there is not other behavior unless access/change a resource which can be done directly through the instance w/o any other method (and we don't use getters/setters in Gossamer)

Reply:
Given @timwu20 explanation the usage of the `type helper struct` is bc we use `uint` for enc/dec compacts so the correct thing to do is improve the scale codec pkg to handle Codec in a more expressive way where you can define the inner type of the Compact, I would suggest to looking into this https://github.com/crypto2lab/scale-codec/blob/bf52f07636d1b805ccb13c0c2f542ce717ccbc9c/compact_test.go#L55 (I did a while ago) then we can express the Header like this

Reply:
> Can we remove the helper and use the method UnmarshalSCALE in the Header? I really dislike we should maintain 2 structs (adds complexity), so looks like we should not be generic over the number type.

The `helper` is a private type that's only used in `MarshalSCALE` and `UnmarshalSCALE`.  The `helper` is only introduced to facilitate the encoding of the block number since we have to be compatible with substrate and they encode the number using compact encoding (see [substrate code](https://github.com/paritytech/parity-scale-codec/blob/f8e228a4212bf7421800d2579f9cc4f623abf59a/src/compact.rs#L89)).  We denote the same encoding as `uint`, so that's why it differs in the `helper` struct.  

> Also the getters/setters are really useful? Looks like there is not other behavior unless access/change a resource which can be done directly through the instance w/o any other method (and we don't use getters/setters in Gossamer)
The getters/setters are necessary since I'm choosing to utilize headers through the `runtime.Header` interface.  I think it's a good choice to make the code flexible enough to deal with any implementation of `runtime.Header` rather than referencing `generic.Header` directly.

Based on our discussion I'll rename it to something more specific 😄